### PR TITLE
Add limit on ClassLoader WeakReference cache

### PR DESCRIPTION
**Pull request by tylerbenson**
## Problem
Having many WeakReferences can cause problems in an application. To reduce the number, we have a cache lookup to allow us to reuse those WeakReference instances. This cache is currently unlimited in size, and only reduced by cleaning references to GC'd ClassLoaders.  Most applications have a limited number of persistent ClassLoaders. Some applications generate them dynamically.  This is where the cache can grow excessively.

## Solution
Add a size limit to the cache so that less commonly used loaders are evicted.

Note: This could cause a common entry to be evicted which may result in a less performant lookup (because identity comparison won't match for subsequently generated WeakReferences). This is considered an acceptable tradeoff to avoid excessive memory use.

Related: #7678

### CODE REVIEW

**Comment by laurit** [comment]
@opentelemetrybot update

**Comment by laurit** [comment]
I measured heap usage for `WeakLockFreeCache` from starting Liferay (osgi application with many class loaders). For this I took heap dump after Liferay had started and in Eclipse memory analyzer looked up all instances of `WeakLockFreeCache` and summed the retained heap column.
baseline `1,012` objects consuming `13,117,752`
this pr  `1,012` objects consuming `18,115,592`
https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/7866 `806` objects consuming `1,017,816`

Surprisingly the bound cache used in this pr does not result in smaller memory usage. As far as I can tell it is because `ConcurrentLinkedHashMap` is fairly large data structure. On my macbook https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/4d202ba7e9a182b8c8de2d4aba0c60722a785be5/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/cache/concurrentlinkedhashmap/ConcurrentLinkedHashMap.java#L232 is a 16x128 array which filled with atomic refs already consumes `41,296`

**Comment by tylerbenson** [comment]
@trask per your request, I created a test that runs with the full javaagent.  When run on main the test fails, when run with this branch or without the javaagent the test passes.  (Notice, I `@Ignore`d it to avoid running it in CI.)

@laurit I'm not familiar with Liferay.  How many classloaders would you say it generates/uses?  If the classloader usage is spread out more it probably isn't as problematic.  This PR is explicitly trying to handle **dynamically** generated classloaders.

**Comment by tylerbenson** [comment]
@laurit Thanks for the pointer about `@Disabled`.

The reason that tracking and keeping `Runtime.getRuntime().totalMemory()` constrained is important is because of systems like Docker that monitor the process memory usage and kill the process if it exceeds a configured threshold.  Minimizing large swings in allocated memory is valuable and often preferred even with a tradeoff of higher average memory (or even cpu) usage.  Controlling those large swings is ultimately the goal of configuring these cache limits.

**Comment by tylerbenson** [comment]
@laurit I ran the included test on `main` with your change included and it does indeed solve the problem.  I will go ahead and close this PR.  Thanks for the fix.

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/MuzzleMatcher.java [hunk 1]
```diff
@@ -39,7 +39,7 @@ class MuzzleMatcher implements AgentBuilder.RawMatcher {
   private final InstrumentationModule instrumentationModule;
   private final Level muzzleLogLevel;
   private final AtomicBoolean initialized = new AtomicBoolean(false);
-  private final Cache<ClassLoader, Boolean> matchCache = Cache.weak();
+  private final Cache<ClassLoader, Boolean> matchCache = Cache.weakBounded(25);
```
Comment (tylerbenson) [inline javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/MuzzleMatcher.java:42]: @laurit do you think your change should be applied to this matcher too?

**Pull request closed without merging**